crypto/threads_common.c

```diff
@@ -0,0 +1,451 @@
+ *
+ * @param arg
+ *        Pointer to the parent `MASTER_KEY_ENTRY` which may contain a
+ *        cleanup function.
+ */
+static void clean_ctx_entry(ossl_uintmax_t idx, CTX_TABLE_ENTRY *ctxentry, void *arg)
+{
+    OPENSSL_free(ctxentry);
+}
+
+/**
+ * @brief Cleans up all context-specific entries for a given key ID.
+ *
+ * This function is used to release all context data associated with a
+ * specific thread-local key (identified by `idx`). It iterates over the
+ * context table in the given `MASTER_KEY_ENTRY`, invoking cleanup for each
+ * `CTX_TABLE_ENTRY`, then frees the context table and the entry itself.
+ *
+ * @param idx
+ *        The key ID associated with the `MASTER_KEY_ENTRY`. Unused.
+ *
+ * @param entry
+ *        Pointer to the `MASTER_KEY_ENTRY` containing the context table
+ *        to be cleaned up.
+ *
+ * @param arg
+ *        Unused parameter.
+ */
+static void clean_master_key_id(ossl_uintmax_t idx, MASTER_KEY_ENTRY *entry, void *arg)
+{
+    ossl_sa_CTX_TABLE_ENTRY_doall_arg(entry->ctx_table, clean_ctx_entry, entry);
+    ossl_sa_CTX_TABLE_ENTRY_free(entry->ctx_table);
+    OPENSSL_free(entry);
+}
+
+/**
+ * @brief Cleans up all master key entries for the current thread.
+ *
+ * This function is the top-level cleanup routine for the thread-local
+ * storage associated with OpenSSL master keys. It is typically registered
+ * as the thread-local storage destructor. It iterates over all
+ * `MASTER_KEY_ENTRY` items in the sparse array, releasing associated
+ * context data and structures.
+ *
+ * @param data
+ *        Pointer to the thread-local `SPARSE_ARRAY_OF(MASTER_KEY_ENTRY)`
+ *        structure to be cleaned up.
+ */
+static void clean_master_key(void *data)
+{
+    SPARSE_ARRAY_OF(MASTER_KEY_ENTRY) *mkey = data;
+
+    ossl_sa_MASTER_KEY_ENTRY_doall_arg(mkey, clean_master_key_id, NULL);
+    ossl_sa_MASTER_KEY_ENTRY_free(mkey);
+}
+
+/**
+ * @brief Initializes the thread-local storage for master key data.
+ *
+ * This function sets up the thread-local key used to store per-thread
+ * master key tables. It also registers the `clean_master_key` function
+ * as the destructor to be called when the thread exits.
+ *
+ * This function is intended to be called once using `CRYPTO_THREAD_run_once`
+ * to ensure thread-safe initialization.
+ */
+static void init_master_key(void)
+{
+    /*
+     * Note: We assign a cleanup function here, which is atypical for
+     * uses of CRYPTO_THREAD_init_local.  This is because, nominally
+     * we expect that the use of ossl_init_thread_start will be used
+     * to notify openssl of exiting threads.  However, in this case
+     * we want the metadata for this interface (the sparse arrays) to
+     * stay valid until the thread actually exits, which is what the
+     * clean_master_key function does.  Data held in the sparse arrays
+     * (that is assigned via CRYPTO_THREAD_set_local_ex), are still expected
+     * to be cleaned via the ossl_init_thread_start/stop api.
```

Comment:
Some optional enhancement in the future, might be to register data specific cleanup handlers. Sometimes `ossl_init_thread_stop` doesn't work and applications can experience leaks (e.g. if OpenSSL is unloaded while threads are still running). At the moment we rely on the user application to call `OPENSSL_thread_stop` for all threads in such a case - which is clunky, not well known or understood by application developers, and prone to errors. This also happens on Windows when static linking because we get no notification of threads exiting there.

Reply:
agreed.  Given that this is an internal-only api we should be able to do so at our leisure in the future